activiti-cloud-services-runtime-bundle/activiti-cloud-services-job-executor/src/main/java/org/activiti/cloud/services/job/executor/DefaultJobMessageProducer.java

```diff
@@ -43,31 +42,57 @@ public DefaultJobMessageProducer(BinderAwareChannelResolver resolver,
+        Assert.hasLength(destination, "destination must not be empty");
+        
+        // Let's try to resolve message channel while inside main Activiti transaction to minimize infrastructure errors 
+        MessageChannel messageChannel = resolver.resolveDestination(destination);
+
+        // Let's send message right after the main transaction has successfully committed. 
+        TransactionSynchronizationManager.registerSynchronization(new JobMessageTransactionSynchronization(destination, 
+                                                                                                           messageChannel, 
+                                                                                                           job));
+    }
+    
+    protected Message<String> buildMessage(Job job) {
+        return MessageBuilder.withPayload(job.getId())
+                // TODO set headers?  
+                .build();
     }
     
-    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
-    public void on(@NonNull JobMessageEvent event) {
-        logger.debug("On JobMessageEvent: {}", event);
+    class JobMessageTransactionSynchronization implements TransactionSynchronization {
 
-        Assert.isTrue(TransactionSynchronizationManager.isActualTransactionActive(), "requires actual active transaction");
+        private final MessageChannel messageChannel;
+        private final Job job;
+        private final String destination;
 
-        Assert.notNull(event.getJobId(), "job id must not be null");
-        Assert.notNull(event.getDestination(), "destination must not be null");
-        
-        Message<String> message = buildMessage(event);
-        
-        // Let's send message 
-        MessageChannel messageChannel = resolver.resolveDestination(event.getDestination());
-        
-        messageChannel.send(message);
-    }
+        public JobMessageTransactionSynchronization(String destination, MessageChannel messageChannel, Job job) {
+            this.destination = destination;
+            this.messageChannel = messageChannel;
+            this.job = job;
+        }
 
-    protected Message<String> buildMessage(JobMessageEvent event) {
-        return MessageBuilder.withPayload(event.getJobId())
-                // TODO set headers?  
-                .build();
-        
+        @Override
+        public void afterCommit() {
+            Message<String> message = buildMessage(job);
+            
+            logger.debug("Sending job message '{}' to destination '{}' via message channel: {}", message, destination, messageChannel);
+            
+            try { 
+                boolean sent = messageChannel.send(message);
+                
+                if(!sent)
+                    throw new RuntimeException("Job message cannot be sent due to non-fatal reason from message channel.");
+
+                eventPublisher.publishEvent(new JobMessageSentEvent(job.getId(), destination, job));
```

Comment:
Who is listening to this event?

Reply:
This is pluggable extension to allow custom application logic to handle job message application events, i.e. logging, error channel DLQ, etc.